internal/c/parts/video/font/font.cpp

```diff
@@ -1,339 +1,1417 @@
+/// @param codepoint The UTF32 codepoint array
+/// @param codepoints The number of codepoints
+/// @return Length in pixels
+int32_t FontPrintWidthUTF32(int32_t fh, const uint32_t *codepoint, int32_t codepoints) {
+    if (codepoints > 0) {
+        FONT_DEBUG_CHECK(IS_FONT_HANDLE_VALID(fh));
+
+        // Get the actual width in pixels
+        return fontManager.fonts[fh]->GetStringPixelWidth((FT_ULong *)codepoint, codepoints);
+    }
+
+    return 0;
+}
+
+/// @brief Returns the length of an ASCII codepoint string in pixels
+/// @param fh A valid font
+/// @param codepoint The ASCII codepoint array
+/// @param codepoints The number of codepoints
+/// @return Length in pixels
+int32_t FontPrintWidthASCII(int32_t fh, const uint8_t *codepoint, int32_t codepoints) {
+    if (codepoints > 0) {
+        FONT_DEBUG_CHECK(IS_FONT_HANDLE_VALID(fh));
+
+        // Atempt to convert the string to UTF32
+        if (utf32.ConvertASCII(codepoint, codepoints)) {
+            // Get the actual width in pixels
+            return FontPrintWidthUTF32(fh, (uint32_t *)utf32.codepoint, utf32.codepoints);
+        }
+    }
+
+    return 0;
+}
+
+/// @brief Master rendering routine (to be called by all other functions). None of the pointer args can be NULL
+/// @param fh A valid font handle
+/// @param codepoint A pointer to an array of UTF-32 codepoints that needs to be rendered
+/// @param codepoints The number of codepoints in the array
+/// @param options 1 = monochrome where black is 0 & white is 255 with nothing in between
+/// @param out_data A pointer to a pointer to the output pixel data (alpha values)
+/// @param out_x A pointer to the output width of the rendered text in pixels
+/// @param out_y A pointer to the output height of the rendered text in pixels
+/// @return success = 1, failure = 0
+bool FontRenderTextUTF32(int32_t fh, const uint32_t *codepoint, int32_t codepoints, int32_t options, uint8_t **out_data, int32_t *out_x, int32_t *out_y) {
+    FONT_DEBUG_CHECK(IS_FONT_HANDLE_VALID(fh));
+
+    auto fnt = fontManager.fonts[fh];
+
+    // Safety
+    *out_data = nullptr;
+    *out_x = 0;
+    *out_y = fnt->defaultHeight;
+
+    if (codepoints <= 0)
+        return codepoints == 0; // true if zero, false if -ve
+
+    auto isMonochrome = options & FONT_RENDER_MONOCHROME; // do we need to do monochrome rendering?
+    FT_Vector strPixSize = {
+        fnt->GetStringPixelWidth((FT_ULong *)codepoint, codepoints), // get the total buffer width
+        fnt->defaultHeight                                           // height is always set by the QB64
+    };
+    auto outBuf = (uint8_t *)calloc(strPixSize.x, strPixSize.y);
+    if (!outBuf)
+        return false;
+
+    FONT_DEBUG_PRINT("Allocated (%llu x %llu) buffer", strPixSize.x, strPixSize.y);
+
+    FT_Pos penX = 0;
+
+    if (fnt->monospaceWidth) {
+        for (FT_ULong i = 0; i < codepoints; i++) {
+            auto cp = codepoint[i];
+
+            if (fnt->CacheGlyph(cp)) {
```

Comment:
It's minor, but rather than expose `CacheGlyph()` and `glyphs` directly I'd just add a `GetGlyph(cp)` that returns the `Glyph` pointer and automatically loads the cache if it's not present. It makes the caching aspect transparent to callers.

Reply:
That's a good idea.

Reply:
This is done.